### PR TITLE
Adding start of work to better clarify links for the image spec

### PR DESCRIPTION
**Pull request by vsoch**
One of our goals for the Containers Working Group is to better clarify symbolic and hard links in the image spec:

https://supercontainers.github.io/containers-wg/ideas/links_B6/

We started discussion in this issue: https://github.com/opencontainers/image-spec/issues/857
And then moved over to the document here: https://docs.google.com/document/d/1uNigsONIndSUbzhtU-2H7zvsgzwrryV3RfJtI_aNxyQ/edit# (please refer to that for some notes that aren't in the changeset here!)

This pull request simply takes the bulk of our discussion and notes and takes a first shot at adding it to the spec document! I hope most of this content would be okay to add given that it's reflecting current practice. Where there is undefined behavior or doubt, I think for the time being we can just state that.

ping @cyphar and @reidpr that contributed most here and have breadth of experience with links! 

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

**Comment by vbatts** [comment]
the discussion over, right? and this is just up for review now?

**Comment by vsoch** [comment]
Yes I believe we finished our discussion on the more living doc (google docs) and this is ready for review!

File: layer.md [hunk 1]
```diff
@@ -186,6 +186,38 @@ rootfs-c9d-v1.s1/
...
 ```
 
+### Symbolic Links
+
+When applying a changeset, (assessing the changeset, or creating the original comparison filesystem) symlinks that are either:
+
+- Already existing in a previous layer
+- Added in the current layer
+
+SHOULD NOT be followed.
```
Comment (vbatts) [inline layer.md:196]: let's move this sentence fragment into the line 191, like "symlinks SHOULD NOT be follow that are either:"

File: layer.md [hunk 2]
```diff
@@ -186,6 +186,38 @@ rootfs-c9d-v1.s1/
...
 
+### Symbolic Links
+
+When applying a changeset, (assessing the changeset, or creating the original comparison filesystem) symlinks that are either:
+
+- Already existing in a previous layer
+- Added in the current layer
+
+SHOULD NOT be followed. 
+If the path of a changeset contains a symlink component, the symlink component MUST be removed and replaced with a directory as well as any necessary subdirectories to reconstruct the path of the changeset.
```
Comment (vbatts) [inline layer.md:197]: with the above change, then this will stand more on its own, which makes it more readable IMO

File: layer.md [hunk 3]
```diff
@@ -186,6 +186,38 @@ rootfs-c9d-v1.s1/
...
+- Already existing in a previous layer
+- Added in the current layer
+
+SHOULD NOT be followed. 
+If the path of a changeset contains a symlink component, the symlink component MUST be removed and replaced with a directory as well as any necessary subdirectories to reconstruct the path of the changeset. 
+If there are changeset entries for any of these directories, the corresponding changeset MUST still be applied on those directories so that they have the correct metadata. If there is not a changeset entry for a directory created in this manner, the metadata of the created directory is implementation-defined but implementations SHOULD strive to make the results consistent. 
+Implementations SHOULD NOT produce changesets which have paths that contain symlink components.
+
+However, an implementation MAY choose to provide a means to allow the user to explicitly ask to follow symlink path components.
+As an example, umoci provides a flag `--keep-dirlinks`.
```
Comment (vbatts) [inline layer.md:202]:
```suggestion
As an example, [umoci](https://github.com/opencontainers/umoci) provides a flag `--keep-dirlinks`. 
```

File: layer.md [hunk 4]
```diff
@@ -186,6 +186,38 @@ rootfs-c9d-v1.s1/
...
+Implementations SHOULD NOT produce changesets which have paths that contain symlink components.
+
+However, an implementation MAY choose to provide a means to allow the user to explicitly ask to follow symlink path components.
+As an example, umoci provides a flag `--keep-dirlinks`. 
+Given such a directive to follow symlinks, such links SHOULD only be followed within the scope of the root of the comparison filesystem.
+
+Implementations SHOULD NOT restrict the target of a symlink, because symlinks are not resolved at linking time and thus adding restrictions on the target of a symlink is not necessary -- the container process will be restricted in its ability to resolve paths outside of the container root filesystem.
+
+Link replacement in a later layer SHOULD be replaced, following the same convention as a changeset over existing files.
```
Comment (vbatts) [inline layer.md:207]: this sentence is not clear to me

**Comment by vbatts** [comment]
Bump for updates and review (covered in the call today)

File: layer.md (hunk 4 above)
Comment (vsoch) [reply layer.md:207]: I think it's saying if we find a link replacement in some layer after one we've parsed, we should honor it. How should we say that so it's clear?

Comment (vsoch) [reply layer.md:207]: It might just be implied this happens and this is too much detail, so I can also take it out. cc @cyphar and @reidpr

**Comment by vsoch** [comment]
@vbatts changes are in! For the last point that I pinged others to check on, I think we can actually remove it because parsing additional layers and handling symlinks is expected to happen. The others can chime in if we should tweak this further.

File: layer.md (hunk 4 above)
Comment (reidpr) [reply layer.md]: It's a little fuzzy for me too but symlinks are so tricky I'm in favor of being explicit, even if in some kind of informal note.

Comment (vsoch) [reply layer.md]: okay I'm not sure what it's supposed to say - maybe @cyphar can add insight.

File: layer.md [hunk 5]
```diff
@@ -186,6 +186,36 @@ rootfs-c9d-v1.s1/
         my-app-tools
 ```
 
+### Symbolic Links
+
+When applying a changeset, (assessing the changeset, or creating the original comparison filesystem) symlinks SHOULD NOT be followed that are either:
+
+- Already existing in a previous layer
+- Added in the current layer
```
Comment (sudo-bmitch) [inline layer.md:194]: Do you have an example that doesn't fit these two scenarios?

**Review by sudo-bmitch — COMMENTED** [review]
I don't fully understand the scenario being handled by the symlink definition. There are quite a few images that include intentional symlinks that we don't want to convert to a copy of the file or directory tree. Will this impact images that link logfiles to `/dev/stdout`, or things like the `/lib/x86_64-linux-gnu` directory contents in a debian image? If not, then I'm not following what scenario is being described by:

> If the path of a changeset contains a symlink component, the symlink component MUST be removed and replaced with a directory as well as any necessary subdirectories to reconstruct the path of the changeset.

**Pull request closed without merging**